content/blog/2019-11-mio-v0.7-alpha.md

```diff
@@ -0,0 +1,201 @@
+now looks like this:
+
+```rust
+poll.registry().register(&socket, Token(0), Interests::READABLE)?;
+```
+
+
+#### Moving to edge triggers
+
+As already mentioned Mio now registers all I/O source with edge triggers which
+means that users of one-shot and level trigger need to change how they respond
+to events.
+
+The problem was that certain usages of level and one-shot triggers showed
+differences in OSes that Mio couldn't, without introducing a lot of overhead,
+smooth over to make Mio a nice cross-platform experience. Instead we decided to
+change all triggers to be edge triggers, which perform the same way on all
+platforms.
+
+Users that previously used one-shot triggers should now `deregister` the I/O
+source if they no longer want to receive events for it. And once more events are
+wanted the resource should be `register`ed again.
+
+Users that used level triggers basically need to put a loop around all I/O
+operations. When using edge triggers the user is responsible for attempting the
+I/O operation, e.g. a `read`, until it returns a would block error (an
+`io::Error` of kind `io::ErrorKind::WouldBlock`). Only once an operation
+returned a would block error the source will receive more events for that kind
+of operation. Below is a sample example of how to read from a `TcpStream` using
+edge triggers.
+
+```rust
+let stream: TcpStream = ...;
+
+// After we polled we received an readable event for our `TcpStream`.
+
+let mut buf = [0; 4096];
+// With edge triggers we need to read all data available on the socket.
+loop {
+    let bytes_read = match stream.read(&mut buf) {
+        // Read successful, proceed like normal.
+        Ok(bytes_read) => bytes_read,
+        // No more data to read at the moment. Now we will receive another event
+        // once more data is available to read.
+        Err(ref err) if err.kind() == io::ErrorKind::WouldBlock => break,
+        // Got interrupted, try again.
+        Err(ref err) if err.kind() == io::ErrorKind::Interrupted => continue,
+        // Hit an actual error.
+        Err(err) => return Err(err),
+    };
+
+    process_byte(&buf[0..bytes_read]);
+}
+```
+
+**Note**: this needs to happen for all I/O operations on registered event source
+so be careful when using automated tools to convert to the new register API.
+
+
+### Addition of Unix socket API
+
+New in the `net` module are the `UnixListener`, `UnixStream` and `UnixDatagram`
+types that support Unix sockets with an API similar to that found in the
+standard library.
+
+These APIs are currently only support on Unix based OSes.
+
+
+### Removal of deprecated API
+
+In Mio version 0.6 various types, such as the old event loop related and channel
+types, were deprecated. In version 0.7 all deprecated type were removed.
+
+
+### Removed support for OSes
+
+Support for the following OSes is dropped:
+
+ * Linux below version 4, we using newer APIs not present on old Linux versions.
```

Comment:
s/, we using/: we are using/ and s/old/older/

Reply:
I thought we were copying libuv on the support front: https://github.com/libuv/libuv/blob/v1.x/SUPPORTED_PLATFORMS.md

Granted, we don't have CI for some of those... so i guess that is "best effort"?

Reply:
I've looked through the Linux code and the manuals and found the following requirements:
* `sys::Waker`: uses eventfd(2), requires Linux 2.6.27, glibc 2.8.
* `sys::net` types: uses socket(2) with SOCK_NONBLOCK and SOCK_CLOEXEC, requires Linux 2.6.27.
* `sys::Selector`: uses epoll_create1(2), requires Linux 2.6.27 and glibc 2.9.

So I would say Linux version 2.6.27 (and glibc 2.9) are required. I don't know why libuv requires 2.6.32 (and glibc 2.12), maybe because RHEL 6 uses 2.6.32?